### DISCUSSION TITLE
Strange Nullable-Analysis results - Tests?

### DISCUSSION THREAD

**Discussion by bernd5** [dotnet/roslyn#51669] [Q&A]
Currently I try to remove redundant null checks based on the flow-state nullability analysis. I want to remove only checks where a variable is guaranteed not null.
Unfortunately the results I get are really strange.
Are there some nullability test cases in the repo which do not bypass the public API surface?

Currently my IsNotNull implementation looks like that:
```CSharp
public static bool IsNotNull(this TypeInfo info)
{
    if (info.Nullability.FlowState == NullableFlowState.NotNull)
    {
        return true;
    }
    return false;
}

public  static bool IsNotNull(this ExpressionSyntax expr, SemanticModel model)
{
    var info = model.GetTypeInfo(expr);
    return IsNotNull(info);
}
```

I found only [this](https://github.com/dotnet/roslyn/blob/main/docs/features/nullable-reference-types.md) document regarding nullability.

**Comment by svick** [comment]
Can you show some of the strange results that you're seeing?

**Reply by bernd5** [reply]
[Here ](https://dotnetfiddle.net/df9eNs) you can find a small sample.

I don't understand why `temp` is `NotNull` in the first switch arm. In my eyes this is wrong - it should be still `MaybeNull`.

**Reply by bernd5** [reply]
The problem seems to be caused by the declared nullability. 
Take for example this code: https://dotnetfiddle.net/eMiZ81

The ToTemp function does not handle nullability at all - it should work for nullable and non-nullable types.

**Reply by RikkiGibson** [reply]
> [Here ](https://dotnetfiddle.net/df9eNs) you can find a small sample.
> 
> I don't understand why `temp` is `NotNull` in the first switch arm. In my eyes this is wrong - it should be still `MaybeNull`.

I'm reading "the first switch arm" as line 54, `_ when temp is {} xxx => ""Something""`.

Here's what's happening:
1. The compiler is inferring the type argument to `ToTemp` based on the inputs. In this case there's just the `string` returned by `Eval(""A"")`. So the type argument is `string`.
    - Note that the `[NotNullIfNotNull]` attribute here isn't particularly necessary since the type argument inference is going to use the flow state of the input argument.
2. Nullable analysis assigns a value of type `string` to the out argument `temp`.
3. `temp` is used on the next line and its flow state is not-null due to having a not-null string assigned to it.

**Reply by bernd5** [reply]
So practially the reason is because of non perfect return type annotation. 
Is there a way to use it in a more pessimistic mode?

Here `Eval` does not do what it says - it returns a null value - and the return type is declared in "classical style" which means now not null - in the past this was not the case (assume it is located in a different file...).
The problem is that the majority of code is written before we had nullable reference types. In an optimization szenario I can't assume that a null-check is not required just because the pure not annotated type is used...

**Reply by RikkiGibson** [reply]
Nullable analysis assumes that method signatures are accurate. When the implementation doesn't meet the signature, e.g. return type is `string` but the method can return `null`, a warning is given in that implementation, not at the call site.

> Is there a way to use it in a more pessimistic mode?

No. If you wanted that, I think you'd be looking at implementing your own flow analysis which is much more pessimistic about when things could be null. Keep in mind that there's always an aspect of practicality to nullable analysis. For example we could pessimistically assume that a certain method not only returns null when it claims to return not-null, but it also nulls out all the fields on the containing type or on the parameters. But analyzing with the assumption that methods may do this even if they do not indicate that they will in their signature doesn't end up being very useful in practice.

**Reply by jasonmalinowski** [reply]
@RikkiGibson: I don't think the problem is the annotation lying, but rather that there isn't an annotation:

> The problem is that the majority of code is written before we had nullable reference types.

So @bernd5 might be getting bit by our handling of oblivious values, right? The annotation of the method being called would be NullableAnnotation.None, but that's converted by the flow analysis engine into NotNull as a way to suppress downstream warnings.
